dvc/repo/__init__.py

```diff
@@ -300,30 +302,25 @@ def graph(self, stages=None, from_directory=None):
         G_active = nx.DiGraph()
         stages = stages or self.stages(from_directory, check_dag=False)
         stages = [stage for stage in stages if stage]
-        outs = []
+        outs = {}
 
         for stage in stages:
             for out in stage.outs:
-                existing = []
-                for o in outs:
-                    if o.path_info == out.path_info:
-                        existing.append(o.stage)
-
-                    in_o_dir = out.path_info.isin(o.path_info)
-                    in_out_dir = o.path_info.isin(out.path_info)
-                    if in_o_dir or in_out_dir:
-                        raise OverlappingOutputPathsError(o, out)
-
-                if existing:
-                    stages = [stage.relpath, existing[0].relpath]
+                if out.path_info in outs:
+                    stages = [stage.relpath, outs[out.path_info].stage.relpath]
                     raise OutputDuplicationError(str(out), stages)
+                outs[out.path_info] = out
 
-                outs.append(out)
+        for stage in stages:
+            for out in stage.outs:
+                for p in out.path_info.parents:
+                    if p in outs:
+                        raise OverlappingOutputPathsError(outs[p], out)
```

Comment:
We could probably make this check, and the one below faster by properly using `map`, `chain` and `set`, and checking intersections, but that would require some Error message changes.